aws/hll/aws-s3-transfer-manager/src/download.rs

```diff
@@ -34,17 +49,178 @@ impl From<GetObjectInputBuilder> for DownloadRequest {
+        self.target_part_size_bytes = cmp::min(size_bytes, MIN_PART_SIZE);
+        self
+    }
+
+    /// Set whether checksum validation is enabled.
+    /// Default is true.
+    pub fn enable_checksum_validation(mut self, enabled: bool) -> Self {
+        self.checksum_validation_enabled = enabled;
+        self
+    }
+
+    /// Set the configuration used by the S3 client
+    pub fn sdk_config(mut self, config: SdkConfig) -> Self {
+        self.sdk_config = Some(config);
+        self
+    }
+
+    /// Set the concurrency level this component is allowed to use. This sets the maximum
+    /// number of concurrent in-flight requests.
+    /// Default is 8.
+    pub fn concurrency(mut self, concurrency: usize) -> Self {
+        self.concurrency = concurrency;
+        self
+    }
+
+    pub fn build(self) -> Downloader {
+        self.into()
+    }
+}
+
+impl From<Builder> for Downloader {
+    fn from(value: Builder) -> Self {
+        let sdk_config = value
+            .sdk_config
+            .unwrap_or_else(|| SdkConfig::builder().build());
+        let client = aws_sdk_s3::Client::new(&sdk_config);
+        Self {
+            target_part_size_bytes: value.target_part_size_bytes,
+            checksum_validation_enabled: value.checksum_validation_enabled,
+            concurrency: value.concurrency,
+            client,
+        }
+    }
+}
+
+/// Download an object in the most efficient way possible by splitting the request into
+/// concurrent requests (e.g. using ranged GET or part number).
+#[derive(Debug, Clone)]
+pub struct Downloader {
+    target_part_size_bytes: u64,
+    checksum_validation_enabled: bool,
+    concurrency: usize,
+    client: aws_sdk_s3::client::Client,
+}
+
+impl Downloader {
+    /// Create a new [Builder]
+    pub fn builder() -> Builder {
+        Builder::new()
+    }
+
+    /// Download a single object from S3.
+    ///
+    /// A single logical request may be split into many concurrent ranged `GetObject` requests
+    /// to improve throughput.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::error::Error;
+    /// use aws_sdk_s3::operation::get_object::builders::GetObjectInputBuilder;
+    /// use aws_s3_transfer_manager::download::{Downloader, DownloadRequest};
+    ///
+    /// async fn get_object(client: Downloader) -> Result<(), Box<dyn Error>> {
+    ///     let request = GetObjectInputBuilder::default()
+    ///         .bucket("my-bucket")
+    ///         .key("my-key")
+    ///         .into();
+    ///
+    ///     let handle = client.download(request).await?;
+    ///     // process data off handle...
+    ///     Ok(())
+    /// }
+    /// ```
+    pub async fn download(&self, req: DownloadRequest) -> Result<DownloadHandle, TransferError> {
+        // if there is a part number then just send the default request
+        if req.input.get_part_number().is_some() {
+            todo!("single part download not implemented")
```

Comment:
Can you expand on what the concern is? I imagine we'll end up with another way to construct a request but it makes sense for Rust SDK users to at least provide some compatibility. 

Note these are internal usage of those types which is inevitable at some point since we're making the request through aws_sdk_s3

Reply:
When we "split" the request, there are some options like [if_match](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/operation/get_object/builders/struct.GetObjectInputBuilder.html#method.if_match) we should copy verbatim into every sub-request.

There are other like [range](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/operation/get_object/builders/struct.GetObjectInputBuilder.html#method.range) that would require special handling: "Oh we can split this range into even smaller ranges, but still need to fall within the original range".

For something like PutObject, there's stuff like [checksum_crc32](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/operation/put_object/builders/struct.PutObjectInputBuilder.html#method.checksum_crc32) where ... if the user passes that for a large object what do we do? If we split it, then the user's checksum won't end up as the whole-object-checksum. Reject the request? Send without splitting it so the user's checksum gets used? Split it anyway, but validate that all the bytes we sent did checksum to that value? 

So... every option needs to be considered. If we do something like: "add special handling for X, Y, and Z options but copy everything else over verbatim". Then what if a new option is added to the model, and it needs special handling, but we didn't catch it and terrible things happen as a result.

I'm not dead-set against using the request options from aws_sdk_s3, but if we have that we should have some allow-list of options that we let through, and reject anything we don't know about

Reply:
Right, the checksum issue is one of the reasons I put a TODO that perhaps we should be using `TryFrom` (fallible version of `From`). 

> I'm not dead-set against using the request options from aws_sdk_s3, but if we have that we should have some allow-list of options that we let through, and reject anything we don't know about

We're definitely going to use these types since at some point we are invoking `aws_sdk_s3` methods requiring them BUT the external request type need not match. You'll note we don't directly take these types, only c convenience for converting from them to `DownloadRequest`. I think we'll offer this kind of convenience for aws-sdk-rust/aws_sdk_s3 users (e.g. `TryFrom<GetObjectFluentBuilder> for DownloadRequest`) but what we expose via FFI of course won't be that. 

I understand the concern and I think right now it's not as clear since we aren't providing an alternative to `GetObjectInputFluentBuilder/GetObjectInput` but I expect we will.